arch: build: interrupt: encode multi level IRQ with C macro magic

ycsin opened the pull request:
This PR encodes the Zephyr multi-level interrupt using C macro magic in the devicetree, rather than depending on the [`gen_defines.py`](https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/dts/gen_defines.py) script during build:

https://github.com/zephyrproject-rtos/zephyr/blob/1fc16e65656e742fbe0adc89e0fdf685c51525a5/scripts/dts/gen_defines.py#L435-L441

https://github.com/zephyrproject-rtos/zephyr/blob/1fc16e65656e742fbe0adc89e0fdf685c51525a5/scripts/dts/gen_defines.py#L456-L469

https://github.com/zephyrproject-rtos/zephyr/blob/1fc16e65656e742fbe0adc89e0fdf685c51525a5/scripts/dts/gen_defines.py#L482-L485

This will resolve the [dt & kconfig cross dependency issue](https://github.com/zephyrproject-rtos/zephyr/pull/63131#issuecomment-1736886469) encountered in #63131, and fixes #62883

After this change, the `*_IRQ_IDX_x_VAL_irq` of a node with an interrupt parent is the local IRQ number of that node (raw value as specified in the devicetree, or ARM GIC IRQ encoded output from `gen_defines.py`), rather than being multi-level encoded. The encoding of the IRQN only happens in certain DT macros, and the right one has to be used for build-time connected ISR, i.e. with `IRQ_CONNECT`

Fixes #62883 

> [!IMPORTANT]
> Blocking #62907

> [!NOTE]  
> **Caveat**: The current approach only works for a maximum of 3 levels, I'm not sure if a while-loop equivalent like the implementation in `gen_defines.py` is possible with the macros, but it is what the Zephyr interrupt encoding scheme currently supports, see [Multi-level Interrupt handling](https://docs.zephyrproject.org/latest/kernel/services/interrupts.html#multi-level-interrupt-handling)

ycsin committed:
devicetree: encode multi-level interrupt number in C devicetree magic

The multi-level encoding of the interrupt number currently
happens in the `gen_defines.py`, which is called in the
`dts.cmake` module after `kconfig.cmake`. However, the number
of bits used by each level is defined in Kconfig and this means
that `gen_defines.py` will not be able to get that information
during build.

To fix this, do the multi-level encoding in C devicetree macro
magic instead of the python script. This ticks one of a
long-standing TODO item from the `gen_defines.py`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt commented on scripts/dts/gen_defines.py:
```diff
@@ -453,20 +453,20 @@ def map_arm_gic_irq_type(irq, irq_num):
-            irq_num |= irq_ctrl.interrupts[0].data["irq"]
-            irq_ctrl = irq_ctrl.interrupts[0].controller
-        return irq_num
+    # def encode_zephyr_multi_level_irq(irq, irq_num):
```
I would remove the commented-out lines. Architecturally speaking, I think this proposal is sound - it's better to do this in C via DT macros than to have part of it live in python.

cfriedt commented on scripts/dts/gen_defines.py:
```diff
@@ -482,7 +482,7 @@ def encode_zephyr_multi_level_irq(irq, irq_num):
                 if "arm,gic" in irq.controller.compats:
                     cell_value = map_arm_gic_irq_type(irq, cell_value)
-                cell_value = encode_zephyr_multi_level_irq(irq, cell_value)
+                # cell_value = encode_zephyr_multi_level_irq(irq, cell_value)
```
Remove commented-out lines, please.

cfriedt commented:
Also worth noting, is that this is actually critical to close residual bugs from using bitmaps to encode a variable number of interrupts per level in a multi-level interrupt context, since the magic number `8` was hard-coded all over the place (python, C, etc).

As is it, the code in `main` is broken without this.

ycsin commented:
Force pushed a documented and cleaner version

ycsin commented:
@galak @mbolivar-ampere @mbolivar could you please take a look?

gmarull commented on include/zephyr/devicetree.h:
```diff
@@ -2409,6 +2409,70 @@
+
+/**
+ * @brief Get the node's Zephyr interrupt number at index
+ * If `CONFIG_MULTI_LEVEL_INTERRUPTS` is enabled, the interrupt number at index will be
```
```suggestion
 * If @kconfig{CONFIG_MULTI_LEVEL_INTERRUPTS} is enabled, the interrupt number at index will be
```

ycsin replied:
fixed

gmarull commented on drivers/gpio/gpio_sifive.c:
```diff
@@ -376,7 +376,7 @@ DEVICE_DT_INST_DEFINE(0,
 
 #define		IRQ_INIT(n)					\
-IRQ_CONNECT(DT_INST_IRQ_BY_IDX(0, n, irq),			\
+IRQ_CONNECT(DT_IRQN_BY_IDX(DT_DRV_INST(0), n),			\
```
usage of `DT_DRV_INST` means there's a missing `INST` version missing

ycsin replied:
you're right, just added a `INST` version for this

gmarull reviewed (dismissed):
haven't reviewed the implementation, just found some general issues. this is also missing tests.

ycsin committed:
tests: devicetree: api: add tests for new DT APIs

Add tests for new DT APIs added in the previous commit:
- `DT_IRQN_BY_IDX`
- `DT_INST_IRQN_BY_IDX`

Added additional tests for the following existing DT APIs when
`CONFIG_MULTI_LEVEL_INTERRUPTS` is enabled:
- `DT_IRQN`
- `DT_INST_IRQN`

Added `qemu_riscv32` for the multi-level interrupt tests.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

jhedberg commented:
Is the assignee correct for this one? @mbolivar-ampere could you please take a look?

mbolivar-ampere reviewed (dismissed):
This is on my queue, just a few minor requests to start out with.

`Z_FOO` is a new naming convention in this header for internal names. I'd rather not add one more. Please stick to the `_INTERNAL` convention if you want to choose names that clearly indicate they're not supposed to be used outside of the header. The `Z_` prefix is the subject of argument elsewhere in naming land as far as I recall.

Second, it looks like this won't work past a certain level of nesting in the interrupt tree, which seems like a concern to me that is not addressed in the commit messages -- what's the story here?

mbolivar-ampere commented:
@cfriedt sorry that I was unclear. `_INTERNAL` is a suffix.

```
% rg _INTERNAL include/zephyr/devicetree*
include/zephyr/devicetree.h
141:#define DT_PATH(...) DT_PATH_INTERNAL(__VA_ARGS__)
3018:	DT_NODE_HAS_STATUS_INTERNAL(node_id, status)
4238:#define DT_PATH_INTERNAL(...) \
4240:/** @brief DT_PATH_INTERNAL() helper: prepends _S_ to a node name
4286:#define DT_NODE_HAS_STATUS_INTERNAL(node_id, status) \
```

ycsin committed:
irq: relocate multi-level irq out of irq.h

Relocate multi-level interrupts APIs out of `irq.h` into
a new file named `irq_multilevel.h` to provide cleaner
separation between typical irq & multilevel ones.

Added preprocessor versions of `irq_to_level_x` as `IRQ_TO_Lx`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
> Second, it looks like this won't work past a certain level of nesting in the interrupt tree, which seems like a concern to me that is not addressed in the commit messages -- what's the story here?

While the `gen_defines.py` method supports the encoding of any number of interrupt levels, other parts of the Zephyr actually only support up to 3rd level:

https://github.com/zephyrproject-rtos/zephyr/blob/90b9809c6f5563c8a8f93310fa65fdb8f321bc06/include/zephyr/irq.h#L283-L307

https://github.com/zephyrproject-rtos/zephyr/blob/90b9809c6f5563c8a8f93310fa65fdb8f321bc06/scripts/build/gen_isr_tables.py#L18-L28

However I just found that the interrupts documentation mentioned :

> _providing support for up to **four** interrupt levels using this arch_

https://github.com/zephyrproject-rtos/zephyr/blob/90b9809c6f5563c8a8f93310fa65fdb8f321bc06/doc/kernel/services/interrupts.rst?plain=1#L78

As far as I can tell, there is only 3 levels, not sure if that's a typo? CC @rgundi

cfriedt commented:
> As far as I can tell, there is only 3 levels, not sure if that's a typo? CC @rgundi 

That seems to be more of a generalization for example purposes. I think there are only 3 levels supported right now in Zephyr.

cfriedt reviewed (changes requested):
If the C macro approach is preferable (and I think it is), can we also delete the python script in this PR?

ycsin commented:
> If the C macro approach is preferable (and I think it is), can we also delete the python script in this PR?

IRQ encoding is just a small part of the `gen_defines.py` script, I think we can convert the `map_arm_gic_irq_type` encoding to C macros _later_ such that IRQ encoding is done solely via C macros and the script will just translate numbers in devicetree as-is to the generated C header

rgundi reviewed (approved):
LGTM

cfriedt commented:
@mbolivar-ampere - would be good to get your stamp on this one.

ycsin commented:
ping devicetree maintainers @mbolivar-ampere @galak

fkokosinski reviewed (approved):
LGTM from the perspective of https://github.com/zephyrproject-rtos/zephyr/pull/62907.

cfriedt commented:
Sorry - I twitched and merged.
